### ISSUE TITLE
Special attributes lead to false positive attribute error with None

### ISSUE THREAD

**Issue by Rapptz** [microsoft/pyright#2053]
**Describe the bug**
Certain special attributes that all objects have are reported as missing when used with `None`.

**To Reproduce**
`None.__class__`

**Expected behavior**
No error.

```py
> python
Python 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> None.__class__
<class 'NoneType'>
```

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/1695103/124353542-1ca4a000-dbd5-11eb-9898-15dc5a75f08b.png)

**VS Code extension or command-line**
Pyright 1.1.154

**Additional context**
This also happens with other attributes that `None` does have, like `__doc__`. I ran into this by using the following code:

```py
def some_routine(a: Optional[bool]) -> None:
    if value not in (True, False, None):
        raise TypeError(f'Invalid class: {a.__class__.__name__}')
    ...
```

Using the type guards would make it work so it narrows down to `Never`:

```py
def some_routine(a: Optional[bool]) -> None:
    if a is not None and not isinstance(a, bool):
        raise TypeError(f'{a.__class__.__name__}')
```

But this would break something simple like:

```py
def class_name(a: Optional[bool]) -> str:
    return f'{a.__class__.__name__}_enhanced'
```

mypy manages to infer these correctly with no error:

```py
reveal_type(None.__class__)
reveal_type(None.__doc__)
# main.py:1: note: Revealed type is "Type[None]"
# main.py:2: note: Revealed type is "Union[builtins.str, None]"
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.155, which I just published. It will also be included in the next release of pylance.

**Comment by Rapptz** [comment]
Hey, thanks a lot for the fix! I've updated and tried checking it out but for some reason some code still reports the error.

```py
None.__class__  # OK

def test(a: Optional[bool]):
    a.__class__ # Error

a = None
a.__class__ #  OK
```

It seems like a union related issue? Either that or I might be misunderstanding something.

**Comment by erictraut** [comment]
Ah, you're correct, there is special processing for the `reportOptionalMemberAccess` diagnostic check that applies only to None within a union. I forgot about that code path. This will be fixed in the next release.
